File: CH6(DP)/maxSubm.py
n, m = map(int, input().split())

a = []

for _ in range(n):
    a.append(list(map(int, input().split())))


if n > m:
    b = [[0] * n for _ in range(m)]

    for i in range(n):
        for j in range(m):
            b[j][i] = a[i][j]

    a = b
    n, m = m, n


answer = float('-inf')


for top in range(n):

    temp = [0] * m

    for bottom in range(top, n):

        for col in range(m):
            temp[col] += a[bottom][col]

        # Kadane
        current = temp[0]
        best = temp[0]

        for col in range(1, m):
            current = max(temp[col], temp[col] + current)
            best = max(best, current)

        answer = max(answer, best)


print(answer)

File: CH6(DP)/maxSubm.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    long long n, m;
    cin >> n >> m;

    vector<vector<long long>> a(n, vector<long long>(m));

    for(long long i = 0; i < n; i++) {
        for(long long j = 0; j < m; j++) {
            cin >> a[i][j];
        }
    }

    if(n > m) {
        vector<vector<long long>> b(m, vector<long long>(n));

        for(long long i = 0; i < n; i++) {
            for(long long j = 0; j < m; j++) {
                b[j][i] = a[i][j];
            }
        }

        a = b;
        swap(n, m);
    }

    long long answer = LLONG_MIN;

    for(long long top = 0; top < n; top++) {

        vector<long long> temp(m, 0);

        for(long long bottom = top; bottom < n; bottom++) {

            for(long long col = 0; col < m; col++) {
                temp[col] += a[bottom][col];
            }

            long long current = temp[0];
            long long best = temp[0];

            for(long long col = 1; col < m; col++) {
                current = max(temp[col], temp[col] + current);
                best = max(best, current);
            }

            answer = max(answer, best);
        }
    }

    cout << answer;
}